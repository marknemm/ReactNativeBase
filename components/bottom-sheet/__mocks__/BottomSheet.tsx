import type { BottomSheetProps } from '@components/bottom-sheet/BottomSheet.interfaces';

const MockBottomSheet = jest.fn().mockImplementation((props: BottomSheetProps) => {
  const ActualComponent = jest.requireActual('@components/bottom-sheet/BottomSheet').default;
  return <ActualComponent {...props} />;
});

export default MockBottomSheet;
