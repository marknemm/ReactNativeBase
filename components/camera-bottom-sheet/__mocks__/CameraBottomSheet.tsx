import type { CameraBottomSheetProps } from '@components/camera-bottom-sheet/CameraBottomSheet.interfaces';

const MockCameraBottomSheet = jest.fn().mockImplementation((props: CameraBottomSheetProps) => {
  const ActualComponent = jest.requireActual('@components/camera-bottom-sheet/CameraBottomSheet').default;
  return <ActualComponent {...props} />;
});

export default MockCameraBottomSheet;
