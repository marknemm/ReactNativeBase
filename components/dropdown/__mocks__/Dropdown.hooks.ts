/**
 * Mocks the useDropdownItems hook.
 */
export const useDropdownItems = jest.fn().mockImplementation((data) => data);
