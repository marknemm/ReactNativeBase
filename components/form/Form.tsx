import { FormProvider } from 'react-hook-form';
import { View } from 'react-native';
import type { FormProps } from './Form.interface';

/**
 * A component that wraps a form and provides form context to its children.
 *
 * @param props The component {@link FormProps}.
 * @returns The {@link Form} component.
 */
const Form: React.FC<FormProps> = ({ children, form, style, ...restViewProps }) => (
  <FormProvider {...form}>
    <View
      style={style}
      testID="rnb-form-view"
      {...restViewProps}
    >
      {children}
    </View>
  </FormProvider>
);

export type * from './Form.interface';
export default Form;
