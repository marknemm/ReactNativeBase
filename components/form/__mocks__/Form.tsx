import type { FormProps } from '@components/form/Form.interface';

const FormMock = jest.fn().mockImplementation((props: FormProps) => {
  const ActualComponent = jest.requireActual('@components/form/Form').default;
  return <ActualComponent {...props} />;
});

export default FormMock;
