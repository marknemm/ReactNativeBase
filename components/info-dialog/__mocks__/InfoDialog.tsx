import type { InfoDialogProps } from '@components/info-dialog/InfoDialog.interfaces';

const MockInfoDialog = jest.fn().mockImplementation((props: InfoDialogProps) => {
  const ActualComponent = jest.requireActual('@components/info-dialog/InfoDialog').default;
  return <ActualComponent {...props} />;
});

export default MockInfoDialog;
