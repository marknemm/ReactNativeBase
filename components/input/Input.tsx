import { useFormControl, useFormErrorMessage, useValidationRules } from '@hooks/form-hooks';
import { useCallbacks } from '@hooks/state-hooks';
import { Input as RneInput, useTheme } from '@rneui/themed';
import { forwardRef, useEffect, useRef, useState } from 'react';
import { Controller } from 'react-hook-form';
import { useMaskedInputProps } from 'react-native-mask-input';
import type { InputFC, InputProps, InputRefType } from './Input.interfaces';
import { useStyles } from './Input.styles';

/**
 * Component for generic user text input.
 *
 * @param props The component {@link InputProps}.
 * @param ref The component reference.
 * @returns The {@link Input} component.
 */
const Input: InputFC = forwardRef((props, ref) => {
  const { label, name, onBlur, onChangeText } = props;

  // Derive entities related to an input controlled by react-hook-form
  const control = useFormControl(props);
  const rules = useValidationRules(props, label?.toString());
  const errorMessage = useFormErrorMessage(rules, props);

  return control
    ? (
      <Controller
        control={control}
        name={name}
        render={({ field: { onChange: onChangeForm, onBlur: onBlurForm, value } }) => (
          <InputControlled
            {...props}
            errorMessage={errorMessage}
            onBlur={(event) => {
              onBlurForm();
              onBlur?.(event);
            }}
            onChangeText={(text) => {
              onChangeForm(text);
              onChangeText?.(text);
            }}
            ref={ref}
            value={value}
          />
        )}
        rules={rules}
      />
    )
    : <InputControlled {...props} ref={ref} />;
});

/**
 * Controlled {@link Input} component.
 *
 * @param props The component {@link InputProps}.
 * @param ref The component reference.
 * @returns The {@link InputControlled} component.
 */
const InputControlled: InputFC = forwardRef(({
  containerStyle,
  inputContainerStyle,
  inputStyle,
  labelStyle,
  mask,
  maxLength,
  maxLengthLimitTyping,
  onChangeText,
  style,
  value,
  ...inputProps
}, ref) => {
  const styles = useStyles({ containerStyle, inputContainerStyle, inputStyle, labelStyle, style });
  const { theme } = useTheme();
  const [uiValue, setUiValue] = useState('');
  const inputRef = useRef<InputRefType>();
  ref ??= inputRef;

  const maskedInputProps = useMaskedInputProps({ mask, onChangeText, value });
  if (mask) {
    onChangeText = maskedInputProps.onChangeText;
    value = maskedInputProps.value;
  }

  const maxLengthNum = (typeof maxLength === 'number')
    ? maxLength
    : maxLength?.value ?? undefined;

  useEffect(() => {
    // Sync value prop with input (UI) value. Done manually to prevent change on each keystroke and prevent lag.
    if (value !== undefined && value !== uiValue) {
      const textInput = (ref as React.MutableRefObject<InputRefType>).current;
      textInput?.setNativeProps({ text: value });
    }
  }, [ref, value, uiValue]);

  return (
    <RneInput
      keyboardAppearance={theme.mode}
      placeholderTextColor={theme.colors.placeholder}
      {...inputProps}
      {...(mask ? maskedInputProps : undefined)}
      containerStyle={styles.container}
      inputContainerStyle={styles.inputContainer}
      inputStyle={styles.input}
      labelStyle={styles.label}
      maxLength={maxLengthLimitTyping ? maxLengthNum : undefined}
      onChangeText={useCallbacks(onChangeText, setUiValue)}
      ref={ref}
      style={styles.style}
      value={undefined} // Do not update value prop directly to prevent input lag (see useEffect above).
    />
  );
});

export type * from './Input.interfaces';
export default Input;
