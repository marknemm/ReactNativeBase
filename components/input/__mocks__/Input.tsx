import type { InputProps } from '@components/input/Input.interfaces';

const MockInput = jest.fn().mockImplementation((props: InputProps) => {
  const ActualComponent = jest.requireActual('@components/input/Input').default;
  return <ActualComponent {...props} />;
});

export default MockInput;
