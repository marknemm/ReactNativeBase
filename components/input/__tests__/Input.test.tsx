import Form from '@components/form/Form';
import Input, { type InputRefType } from '@components/input/Input';
import { genMockForm } from '@hooks/__mocks__/form-hooks';
import AppProvider from '@test/contexts/app/AppProvider';
import { act, fireEvent, render, renderHook, screen, userEvent } from '@testing-library/react-native';
import React, { cloneElement, useRef } from 'react';
import { type NativeSyntheticEvent, type TextInputChangeEventData } from 'react-native';
import { Masks } from 'react-native-mask-input';
import type { DeepPartial } from 'utility-types';

describe('<Input />', () => {
  const label = 'Test Label';
  const placeholder = 'Test Placeholder';
  const typeStr = 'Test Typing';
  const value = 'Test Value';

  describe('value changes', () => {
    it('calls the `onChange` callback when input is typed in', async () => {
      const onChange = jest.fn();
      render(
        <Input onChange={onChange} />,
        { wrapper: AppProvider }
      );

      const textInput = screen.getByTestId('RNE__Input__text-input');
      await userEvent.type(textInput, typeStr);

      for (let i = 0; i < typeStr.length; i++) {
        expect(onChange).toHaveBeenNthCalledWith(i + 1,
          expect.objectContaining<DeepPartial<NativeSyntheticEvent<TextInputChangeEventData>>>({
            nativeEvent: expect.objectContaining({ text: typeStr.slice(0, i + 1) }),
          })
        );
      }
    });

    it('calls the `onChangeText` callback when input is typed in', async () => {
      const onChangeText = jest.fn();
      render(
        <Input
          label={label}
          placeholder={placeholder}
          onChangeText={onChangeText}
        />,
        { wrapper: AppProvider }
      );

      const textInput = screen.getByTestId('RNE__Input__text-input');
      await userEvent.type(textInput, typeStr);

      for (let i = 0; i < typeStr.length; i++) {
        expect(onChangeText).toHaveBeenNthCalledWith(i + 1, typeStr.slice(0, i + 1));
      }
    });

    it('updates the input value when `value` prop changes', () => {
      const ref = renderHook(() => useRef<InputRefType>()).result.current;
      render(
        <Input ref={ref} />,
        { wrapper: AppProvider }
      );

      jest.spyOn(ref.current, 'setNativeProps');
      screen.rerender(
        <Input
          ref={ref}
          value={value}
        />
      );

      expect(ref.current.setNativeProps).toHaveBeenCalledWith({ text: value });
    });

    it('does not update the input value when changed via UI', async () => {
      const ref = renderHook(() => useRef<InputRefType>()).result.current;
      render(
        <Input ref={ref} />,
        { wrapper: AppProvider }
      );

      const textInput = screen.getByTestId('RNE__Input__text-input');
      await userEvent.type(textInput, typeStr);

      jest.spyOn(ref.current, 'setNativeProps');
      screen.rerender(
        <Input ref={ref} />
      );

      expect(ref.current.setNativeProps).not.toHaveBeenCalled();
    });

    it('does not update the input value when value is set to current UI value', async () => {
      const ref = renderHook(() => useRef<InputRefType>()).result.current;
      render(
        <Input ref={ref} />,
        { wrapper: AppProvider }
      );

      const textInput = screen.getByTestId('RNE__Input__text-input');
      await userEvent.type(textInput, typeStr);

      jest.spyOn(ref.current, 'setNativeProps');
      screen.rerender(
        <Input
          ref={ref}
          value={typeStr}
        />
      );

      expect(ref.current.setNativeProps).not.toHaveBeenCalled();
    });
  });

  describe('form control', () => {
    it('updates the form control when the input is typed in', async () => {
      const form = genMockForm({
        input: '',
      });
      const onChangeText = jest.fn();

      render(
        <Form form={form}>
          <Input
            label={label}
            name="input"
            onChangeText={onChangeText}
          />
        </Form>,
        { wrapper: AppProvider }
      );

      const textInput = screen.getByTestId('RNE__Input__text-input');
      await userEvent.type(textInput, typeStr);

      expect(form.getValues('input')).toBe(typeStr);
      expect(onChangeText).toHaveBeenCalledWith(typeStr);
    });

    it('blurs the form control when the input is blurred', async () => {
      const form = genMockForm({
        input: '',
      });
      const onBlur = jest.fn();

      render(
        <Form form={form}>
          <Input
            label={label}
            name="input"
            onBlur={onBlur}
          />
        </Form>,
        { wrapper: AppProvider }
      );

      const textInput = screen.getByTestId('RNE__Input__text-input');
      await userEvent.type(textInput, typeStr);
      fireEvent(textInput, 'blur');

      expect(onBlur).toHaveBeenCalled();
    });
  });

  describe('validation and constraints', () => {
    it('validates the required constraint when form is submitted', async () => {
      const form = genMockForm({
        input: '',
      });

      const renderJSX = (
        <Form form={form}>
          <Input
            label={label}
            name="input"
            required
          />
        </Form>
      );

      render(
        renderJSX,
        { wrapper: AppProvider }
      );

      await act(() =>
        form.handleSubmit(() => {})()
      );

      expect(form.formState.errors.input.type).toBe('required');

      screen.rerender(
        cloneElement(renderJSX, { form })
      );

      const errorText = screen.getByText(`${label} is required`);
      expect(errorText).toBeVisible();
    });

    it('validates max length constraint when form is submitted but typing is not restricted', async () => {
      const errorMsg = 'Max length exceeded';
      const onChangeText = jest.fn();
      const form = genMockForm({
        input: '',
      });

      const renderJSX = (
        <Form form={form}>
          <Input
            label={label}
            maxLength={{ value: 5, message: errorMsg }}
            name="input"
            onChangeText={onChangeText}
          />
        </Form>
      );

      render(
        renderJSX,
        { wrapper: AppProvider }
      );

      const textInput = screen.getByTestId('RNE__Input__text-input');
      await userEvent.type(textInput, '123456');

      await act(() =>
        form.handleSubmit(() => {})()
      );

      expect(form.formState.errors.input.type).toBe('maxLength');

      screen.rerender(
        cloneElement(renderJSX, { form })
      );

      const errorText = screen.getByText(errorMsg);
      expect(errorText).toBeVisible();
      expect(onChangeText).toHaveBeenCalledWith('123456');
    });

    it('Enforces max length constraint when typing', async () => {
      const onChangeText = jest.fn();
      const form = genMockForm({
        input: '',
      });

      render(
        <Form form={form}>
          <Input
            label={label}
            maxLength={5}
            maxLengthLimitTyping
            name="input"
            onChangeText={onChangeText}
          />
        </Form>,
        { wrapper: AppProvider }
      );

      const textInput = screen.getByTestId('RNE__Input__text-input');
      await userEvent.type(textInput, '123456');

      expect(onChangeText).toHaveBeenCalledWith('12345');
      expect(onChangeText).not.toHaveBeenCalledWith('1234567');
    });
  });

  describe('mask', () => {
    it('applies the mask to the input', async () => {
      const form = genMockForm({
        input: '',
      });

      render(
        <Form form={form}>
          <Input
            label={label}
            mask={Masks.USA_PHONE}
            name="input"
          />
        </Form>,
        { wrapper: AppProvider }
      );

      const textInput = screen.getByTestId('RNE__Input__text-input');
      await userEvent.type(textInput, '2234567890');

      expect(form.getValues('input')).toBe('(223) 456-7890');
    });
  });

  describe('snapshots', () => {
    it('renders correctly', () => {
      render(
        <Input />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with label and placeholder', () => {
      render(
        <Input
          label={label}
          placeholder={placeholder}
        />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with label and mask', () => {
      render(
        <Input
          label={label}
          mask={Masks.USA_PHONE}
        />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with error', async () => {
      const form = genMockForm({
        input: '',
      });

      const renderJSX = (
        <Form form={form}>
          <Input
            label={label}
            name="input"
            required
          />
        </Form>
      );

      render(
        renderJSX,
        { wrapper: AppProvider }
      );

      await act(() =>
        form.handleSubmit(() => {})()
      );

      screen.rerender(
        cloneElement(renderJSX, { form })
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
