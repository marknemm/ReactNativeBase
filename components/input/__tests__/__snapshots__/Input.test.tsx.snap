// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> snapshots renders correctly 1`] = `
<View
  style={
    {
      "marginBottom": 6.285714285714286,
      "paddingHorizontal": 0,
      "width": "100%",
    }
  }
  testID="RNE__Input__view-wrapper"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#bbb",
        "borderBottomWidth": 1,
        "borderColor": "#86939e",
        "flexDirection": "row",
        "height": 98.9730612244898,
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <TextInput
      editable={true}
      keyboardAppearance="light"
      onChangeText={[Function]}
      placeholderTextColor="#86939e"
      style={
        {
          "color": "#242424",
          "flex": 1,
          "fontSize": 23.57142857142857,
          "height": 98.9730612244898,
          "minHeight": 40,
        }
      }
      testID="RNE__Input__text-input"
      underlineColorAndroid="transparent"
    />
  </View>
  <Text
    style={
      {
        "color": "#ff3b30",
        "fontSize": 12,
        "margin": 5,
      }
    }
  />
</View>
`;

exports[`<Input /> snapshots renders correctly with error 1`] = `
<View
  testID="rnb-form-view"
>
  <View
    style={
      {
        "marginBottom": 6.285714285714286,
        "paddingHorizontal": 0,
        "width": "100%",
      }
    }
    testID="RNE__Input__view-wrapper"
  >
    <Text
      style={
        {
          "color": "#86939e",
          "fontSize": 23.57142857142857,
          "fontWeight": "bold",
        }
      }
    >
      Test Label
    </Text>
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "height": 98.9730612244898,
          "transform": [
            {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        editable={true}
        keyboardAppearance="light"
        name="input"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholderTextColor="#86939e"
        required={true}
        style={
          {
            "color": "#242424",
            "flex": 1,
            "fontSize": 23.57142857142857,
            "height": 98.9730612244898,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
      />
    </View>
    <Text
      style={
        {
          "color": "#ff3b30",
          "fontSize": 12,
          "margin": 5,
        }
      }
    >
      Test Label is required
    </Text>
  </View>
</View>
`;

exports[`<Input /> snapshots renders correctly with label and mask 1`] = `
<View
  style={
    {
      "marginBottom": 6.285714285714286,
      "paddingHorizontal": 0,
      "width": "100%",
    }
  }
  testID="RNE__Input__view-wrapper"
>
  <Text
    style={
      {
        "color": "#86939e",
        "fontSize": 23.57142857142857,
        "fontWeight": "bold",
      }
    }
  >
    Test Label
  </Text>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#bbb",
        "borderBottomWidth": 1,
        "borderColor": "#86939e",
        "flexDirection": "row",
        "height": 98.9730612244898,
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <TextInput
      editable={true}
      keyboardAppearance="light"
      onChangeText={[Function]}
      placeholder="(___) ___-____"
      placeholderTextColor="#86939e"
      style={
        {
          "color": "#242424",
          "flex": 1,
          "fontSize": 23.57142857142857,
          "height": 98.9730612244898,
          "minHeight": 40,
        }
      }
      testID="RNE__Input__text-input"
      underlineColorAndroid="transparent"
    />
  </View>
  <Text
    style={
      {
        "color": "#ff3b30",
        "fontSize": 12,
        "margin": 5,
      }
    }
  />
</View>
`;

exports[`<Input /> snapshots renders correctly with label and placeholder 1`] = `
<View
  style={
    {
      "marginBottom": 6.285714285714286,
      "paddingHorizontal": 0,
      "width": "100%",
    }
  }
  testID="RNE__Input__view-wrapper"
>
  <Text
    style={
      {
        "color": "#86939e",
        "fontSize": 23.57142857142857,
        "fontWeight": "bold",
      }
    }
  >
    Test Label
  </Text>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#bbb",
        "borderBottomWidth": 1,
        "borderColor": "#86939e",
        "flexDirection": "row",
        "height": 98.9730612244898,
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <TextInput
      editable={true}
      keyboardAppearance="light"
      onChangeText={[Function]}
      placeholder="Test Placeholder"
      placeholderTextColor="#86939e"
      style={
        {
          "color": "#242424",
          "flex": 1,
          "fontSize": 23.57142857142857,
          "height": 98.9730612244898,
          "minHeight": 40,
        }
      }
      testID="RNE__Input__text-input"
      underlineColorAndroid="transparent"
    />
  </View>
  <Text
    style={
      {
        "color": "#ff3b30",
        "fontSize": 12,
        "margin": 5,
      }
    }
  />
</View>
`;
